File: src/main/java/plazma/kernel/lib/sys/Parameters.java
/*
 * Copyright (C) 2012-2023 Oleh Hapon ohapon@users.sourceforge.net
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
 * 
 * Oleh Hapon
 * Kyiv, UKRAINE
 * ohapon@users.sourceforge.net
 */

package plazma.kernel.lib.sys;

import java.util.ArrayList;
import java.util.List;

import plazma.kernel.lib.num.NumLib;

public class Parameters {
    
    private List<Parameter> parameters;

    public Parameters() {
        super();
    }

    public void addParameter(Parameter parameter) {
        if (parameter == null) {
            return;
        }

        getParameters().add(parameter);

    }

    public void addParameter(String name) {
        addParameter(new Parameter(name));
    }

    public void addParameter(String name, String description) {
        addParameter(new Parameter(name, description));
    }

    public void addParameter(String name, String description, String datatType) {
        addParameter(new Parameter(name, description, datatType));
    }

    public void addParameter(String name, String description, String datatType, boolean required) {
        addParameter(new Parameter(name, description, datatType, required));
    }

    public void addParameter(String name, String description, boolean required) {
        addParameter(new Parameter(name, description, required));
    }

    public void addParameter(String name, boolean required) {
        addParameter(new Parameter(name, required));
    }

    ////

    public List<Parameter> getParameters() {
        if (parameters == null) {
            parameters = new ArrayList<Parameter>();
        }
        return parameters;
    }

    public boolean hasParameters() {
        return getParameterCount() > 0;
    }

    public int getParameterCount() {
        return parameters == null ? 0 : parameters.size();
    }

    // byIndex
    public Parameter getParameter(int index) {
        if (parameters == null) {
            return null;
        }
        if (index < 0 && index >= parameters.size()) {
            // Not found
            return null;
        }
        return parameters.get(index);
    }

    // byName
    public Parameter getParameter(String name) {
        if (parameters == null) {
            return null;
        }
        if (name == null || name.isEmpty()) {
            // Not found
            return null;
        }
        int count = getParameterCount();
        Parameter parameter = null;
        for (int i = 0; i < count; i++) {
            parameter = getParameter(i);
            if (equalsName(parameter, name)) {
                return parameter;
            }
        }
        // Not found
        return null;
    }

    public boolean hasParameter(String name) {
        Parameter parameter = getParameter(name);
        return parameter != null;
    }

    public boolean equalsName(Parameter parameter, String name) {
       if (parameter == null || name.isEmpty()) {
           return false;
        }
        return name == parameter.getName(); // TODO: Check 'longName'
    }    

    ////

    // string

    public String getStringParameter(String name) {
        Parameter parameter = getParameter(name);
        if (parameter == null) {
            return null;
        }
        return parameter.getValue();
    }

    public String getStringParameter(String name, String defValue) {
        Parameter parameter = getParameter(name);
        if (parameter == null) {
            return defValue;
        }
        return parameter.getValue();
    }

    // int

    public int getIntParameter(String name) {
        Parameter parameter = getParameter(name);
        if (parameter == null) {
            return 0;
        }
        return NumLib.toInt(parameter.getValue());
    }

    public int getIntParameter(String name, int defValue) {
        Parameter parameter = getParameter(name);
        if (parameter == null) {
            return defValue;
        }
        return NumLib.toInt(parameter.getValue(), defValue);
    }

    // float

    public float getFloatParameter(String name) {
        Parameter parameter = getParameter(name);
        if (parameter == null) {
            return 0.0f;
        }
        return NumLib.toFloat(parameter.getValue());
    }

    public float getFloatParameter(String name, float defValue) {
        Parameter parameter = getParameter(name);
        if (parameter == null) {
            return defValue;
        }
        return NumLib.toFloat(parameter.getValue(), defValue);
    }

    // boolean


    // [UTIL]
    public boolean toBoolean(String value, boolean defValue) {
        if (value == null || value.isEmpty()) {
            return defValue;
        }

        String str = value.toLowerCase();

        // true
        if (str == "true" || str == "yes" || str == "t" || str == "y") {
            return true;
        }

        // false
        if (str == "false" || str == "no" || str == "f" || str == "n") {
            return false;
        }

        // default
        return defValue;

    }

    public boolean toBoolean(String value) {
        return toBoolean(value, false);
    }

    public boolean getBooleanParameter(String name) {
        Parameter parameter = getParameter(name);
        if (parameter == null) {
            return false;
        }
        return toBoolean(parameter.getValue());
    }

    public boolean getBooleanParameter(String name, boolean defValue) {
        Parameter parameter = getParameter(name);
        if (parameter == null) {
            return defValue;
        }
        return toBoolean(parameter.getValue(), defValue);
    }

}

File: src/main/cpp/plazma/lib/sys/Parameters.cpp

#include "plazma/lib/num/numlib.h"
#include "Parameters.h"
#include <algorithm> // WIN32: std::transform

//#include <iostream>

namespace sys {

    Parameters::Parameters() {
        parameters = nullptr;
    }

    Parameters::~Parameters() {
        //std::cout << "Destroy Parameters: Start" << std::endl;
        if (parameters != nullptr) {
            //std::cout << "Destroy Parameters: Point-1" << std::endl;
            int count = parameters->size();
            Parameter* parameter = nullptr;
            //std::cout << "Destroy Parameters: Point-2" << std::endl;
            for (int i = 0 ; i < count; i++) {
                parameter = parameters->at(i);
                //std::cout << "Destroy Parameters: Point-3-" << std::to_string(i) << std::endl;
                if (parameter == nullptr) {
                    continue;
                }
                delete parameter;
            }
        }
        //std::cout << "Destroy Parameters: End" << std::endl;
    }

    ///

    void Parameters::addParameter(Parameter* parameter) {
        if (parameter == nullptr) {
            return;
        }

        //(*getParameters())[parameter->getName()] = parameter;
        getParameters()->push_back(parameter);

    }

    void Parameters::addParameter(const std::string& name) {
        addParameter(new Parameter(name));
    }

    void Parameters::addParameter(const std::string& name, const std::string& description) {
        addParameter(new Parameter(name, description));
    }

    void Parameters::addParameter(const std::string& name, const std::string& description, const std::string& datatType) {
        addParameter(new Parameter(name, description, datatType));
    }

    void Parameters::addParameter(const std::string& name, const std::string& description, const std::string& datatType, bool required) {
        addParameter(new Parameter(name, description, datatType, required));
    }

    void Parameters::addParameter(const std::string& name, const std::string& description, bool required) {
        addParameter(new Parameter(name, description, required));
    }

    void Parameters::addParameter(const std::string& name, bool required) {
        addParameter(new Parameter(name, required));
    }

    ////

    std::vector<Parameter*>* Parameters::getParameters() {
        if (parameters == nullptr) {
            parameters = new std::vector<Parameter*>();
        }
        return parameters;
    }

    bool Parameters::hasParameters() {
        return getParameterCount() > 0;
    }

    int Parameters::getParameterCount() {
        return parameters == nullptr ? 0 : parameters->size();
    }


    // byIndex
    Parameter* Parameters::getParameter(const int index) {
        if (parameters == nullptr) {
            return nullptr;
        }
        if (index < 0 && index >= parameters->size()) {
            // Not found
            return nullptr;
        }
        return parameters->at(index);
    }

    // byName
    Parameter* Parameters::getParameter(const std::string& name) {
        if (parameters == nullptr) {
            return nullptr;
        }
        if (name.empty()) {
            // Not found
            return nullptr;
        }
        int count = getParameterCount();
        Parameter* parameter = nullptr;
        for (int i = 0; i < count; i++) {
            parameter = getParameter(i);
            if (equalsName(parameter, name)) {
                return parameter;
            }
        }
        // Not found
        return nullptr;
    }

    bool Parameters::hasParameter(const std::string& name) {
        Parameter* parameter = getParameter(name);
        return parameter != nullptr;
    }

    bool Parameters::equalsName(Parameter* parameter, const std::string& name) {
       if (parameter == nullptr || name.empty()) {
           return false;
        }
        return name == parameter->getName(); // TODO: Check 'longName'
    }

    ////

    // string

    std::string Parameters::getStringParameter(const std::string& name) {
        Parameter* parameter = getParameter(name);
        if (parameter == nullptr) {
            return "";
        }
        return parameter->getValue();
    }

    std::string Parameters::getStringParameter(const std::string& name, std::string defValue) {
        Parameter* parameter = getParameter(name);
        if (parameter == nullptr) {
            return defValue;
        }
        return parameter->getValue();
    }

    // int

    int Parameters::getIntParameter(const std::string& name) {
        Parameter* parameter = getParameter(name);
        if (parameter == nullptr) {
            return 0;
        }
        return numlib::toInt(parameter->getValue());
    }

    int Parameters::getIntParameter(const std::string& name, int defValue) {
        Parameter* parameter = getParameter(name);
        if (parameter == nullptr) {
            return defValue;
        }
        return numlib::toInt(parameter->getValue(), defValue);
    }

    // float

    float Parameters::getFloatParameter(const std::string& name) {
        Parameter* parameter = getParameter(name);
        if (parameter == nullptr) {
            return 0;
        }
        return numlib::toFloat(parameter->getValue());
    }

    float Parameters::getFloatParameter(const std::string& name, float defValue) {
        Parameter* parameter = getParameter(name);
        if (parameter == nullptr) {
            return defValue;
        }
        return numlib::toFloat(parameter->getValue(), defValue);
    }

    // boolean


    // [UTIL]
    bool toBoolean(const std::string& value, bool defValue) {
        if (value.empty()) {
            return defValue;
        }

        std::string str = value;
        std::transform(str.begin(), str.end(), str.begin(), ::tolower);

        // true
        if (str == "true" || str == "yes" || str == "t" || str == "y") {
            return true;
        }

        // false
        if (str == "false" || str == "no" || str == "f" || str == "n") {
            return false;
        }

        // default
        return defValue;

    }

    bool toBoolean(const std::string& value) {
        return toBoolean(value, false);
    }

    bool Parameters::getBooleanParameter(const std::string& name) {
        Parameter* parameter = getParameter(name);
        if (parameter == nullptr) {
            return false;
        }
        return toBoolean(parameter->getValue());
    }

    bool Parameters::getBooleanParameter(const std::string& name, bool defValue) {
        Parameter* parameter = getParameter(name);
        if (parameter == nullptr) {
            return defValue;
        }
        return toBoolean(parameter->getValue(), defValue);
    }

}
